Refactor /docs into Concepts

JasonAtClockwork opened the pull request:
# Description of Changes

Refactoring all sections to match a concept based approach to the docs.

Closes: #3840 

- Modules - Closes: #3841 
- Databases - Closes: #3842 
- Tables - Closes: #3843 
- Reducers - Closes: #3844 
- Procedures - Closes: #3845
- Views - Closes: #3846 
- Codegen - Closes: #3847 
- SDK API - Closes: #3848 
- Subscriptions - Closes: #3849 - No work done here just moved the current order
- Auth - Closes: #3850  - No work done here just moved the current order

# API and ABI breaking changes

N/A

# Expected complexity level and risk

1

# Testing

- [x] Walking through all code references before pushing

JasonAtClockwork committed:
Added concepts for modules, databases, tables, and reducers

gefjon commented on docs/docs/02200-tables/00100-overview.md:
```diff
@@ -0,0 +1,128 @@
+
+# Tables
+
+Tables are the primary way to store data in SpacetimeDB. All data in SpacetimeDB is stored in memory for extremely low latency and high throughput access. SpacetimeDB also automatically persists all data to disk.
```
"primary" feels odd here, as it implies there are other (secondary) ways to store data, which there aren't.

gefjon commented on docs/docs/02200-tables/00400-access-permissions.md:
```diff
@@ -0,0 +1,158 @@
+
+SpacetimeDB enforces different levels of table access depending on the context. All contexts access tables through `ctx.db`, but the available operations differ based on whether the context is read-write or read-only.
+
+## Reducers - Read-Write Access
```
We should either amend this or add a section for read-write access from procedure `with_tx` transactions. I'll come back later and leave a suggestion, or push a commit, or something.

gefjon reviewed (commented):
Not a thorough review yet, just a few thoughts. Structurally this looks great, thanks so much!

JasonAtClockwork committed:
Update docs/docs/02000-modules/00400-quick-reference.md

Co-authored-by: Phoebe Goldman <phoebe@clockworklabs.io>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Updated to fix notes/comments and add the start for views

JasonAtClockwork committed:
Missing character

JasonAtClockwork commented on docs/docs/02000-modules/00100-overview.md:
```diff
@@ -0,0 +1,69 @@
+- [Create a new module project](/new-modules/creating)
+- [Build and publish your module](/new-modules/building-publishing)
+
+For interactive development with hot-reloading, see [Interactive Development with `spacetime dev`](/new-modules/creating).
```
Yeah I was trying to link directly to the interactive but it wasn't happen during build I've removed this.

JasonAtClockwork committed:
Updated code references for access permissions and views

JasonAtClockwork committed:
Overhaul of the layout of the docs to streamline it

JasonAtClockwork committed:
Refactor Getting Started, clearing up links, removing quick-reference for a cheat sheet

JasonAtClockwork committed:
Final pages, review, and edits

JasonAtClockwork committed:
Small link issue

JasonAtClockwork committed:
Fix for quickstart tests

cloutiertyler commented on docs/docs/00100-getting-started/00100-intro/00500-faq.md:
```diff
@@ -0,0 +1,21 @@
+# FAQ
+
+1. What is SpacetimeDB?
+   It's a cloud platform within a database that's fast enough to run real-time games.
```
```suggestion
   It's a cloud platform within a database that's fast enough to run real-time applications.
```

cloutiertyler commented on docs/docs/00200-modules/00200-creating-modules.md:
```diff
@@ -0,0 +1,210 @@
+The fastest way to get started developing a module is with `spacetime dev`. This interactive command guides you through creating a new SpacetimeDB project with hot-reloading - whenever you save changes to your module code, SpacetimeDB automatically rebuilds and republishes your module.
+
+:::caution
+`spacetime dev` is currently UNSTABLE and subject to breaking changes.
```
Just making it less alarmist

cloutiertyler commented on docs/docs/00200-modules/00200-creating-modules.md:
```diff
@@ -0,0 +1,210 @@
+- Watches your source files for changes
+- Automatically rebuilds and republishes when you save changes
+
+Your database will be available at `http://localhost:3000`.
```
```suggestion
Your database will be available at `https://maincloud.spacetimedb.com`.
```

`spacetime dev` publishes to maincloud by default.

cloutiertyler commented on docs/docs/00200-modules/00400-row-level-security.md:
I think we should probably just remove RLS since it's being subsumed by View functions. However, we can do that in a separate PR.

JasonAtClockwork committed:
Updated TypeScript quickstart:
- Removal of confusing Rust/C# module note
- Added missing useReducer import
- Moved PrettyMessages below users to fix a bug later in the tutorial
- Updated PrettyMessages to have null safety if the user follows the module bindings they may have an error due to spacetime call send_message

JasonAtClockwork committed:
Update docs/docs/00100-getting-started/00100-intro/00400-key-architecture.md

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Update docs/docs/00100-getting-started/00100-intro/00500-faq.md

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Update docs/docs/00200-modules/00200-creating-modules.md

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Update docs/docs/00200-modules/00200-creating-modules.md

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

cloutiertyler reviewed (approved):
I left comments inline. I think this is an excellent first pass. I think once the comments are resolved this can merge. I took a look at the styling too by running it locally. We'll want @randyclockwork to take a pass at some point, but I think this is fine for now.

I would like to take a second pass at this where we really try to distill down the basic concepts and put them together towards the top of the document. As it stands I think the table of contents is a little verbose before getting to the core concepts:

<img width="288" height="905" alt="Image" src="https://github.com/user-attachments/assets/cd0d3226-afe9-4a86-89fd-3d8b508fbcb9" />

I would love to see the table of contents look more like:

```sh
Intro
    Getting Started
    Quickstarts
    Tutorials
Core Concepts
    Databases # I think we modify this to include modules as a section
    Tables
    Functions
    Subscriptions
    Authentication
```

Where if you expand all the sections it would look like:

```sh
Intro
    Getting Started
    Quickstarts
        React
        Vue
        HTML
        Unity
        Unreal
        Godot
        Rust
        ... etc
    Tutorials
        Chat App # With language selectors for all the languages
        Blackholio (Unity)
        Blackholio (Unreal)
Core Concepts
    Databases
        The Database Module
        Developing databases
        Building and publishing
        Transactions & Atomicity # discussion about atomic transactions and what they are (could also go in Functions)
        Automatic migrations # these could also go into an "Advanced usage section"
        Incremental migrations # these could also go into an "Advanced usage section"        
    Functions
        Overview
        Reducers
        Procedures
        Views
    Tables
        Defining tables
        Reading data
        Writing data
        Scheduled tables
        Performance best practices
        Advanced
            System tables
    Subscriptions
        Subscribing to tables
        Reducer callbacks # will become event tables
        Advanced semantics # should perhaps be embedded in one of the above
    Authentication
        SpacetimeAuth
        Clerk
        Auth0
        # ... etc
Other features
        PostgreSQL compatibility   
```

We might at some point need deeper levels of nesting and will need UI design for that.

Also please remove the "old files" before merging.

JasonAtClockwork committed:
Update docs/docs/00400-tables/00600-performance.md

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Update docs/docs/00500-functions/00100-overview.md

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Update docs/docs/00700-client-sdk-languages/00200-codegen.md

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Update docs/docs/00700-client-sdk-languages/00200-codegen.md

Co-authored-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>
Signed-off-by: Jason Larabie <jason@clockworklabs.io>

JasonAtClockwork committed:
Added security note for scheduled tables and added missing sender_auth to ReducerContext doc

JasonAtClockwork committed:
Update new generate-cli-docs.mjs to new cli location
